[JSC] Add a resolve_and_get_from_scope bytecode opcode

hyjorc1 opened the pull request:
#### d7468b46b1d24a910392a1b23aec25f0e96e8f87
<pre>
[JSC] Add a resolve_and_get_from_scope bytecode opcode
<a href="https://bugs.webkit.org/show_bug.cgi?id=237836">https://bugs.webkit.org/show_bug.cgi?id=237836</a>

Reviewed by NOBODY (OOPS!).

In generated bytecodes, resolve_scope followed by get_from_scope is common.
This patch introduces a new bytecode opcode resolve_and_get_from_scope by
replacing the resolve_scope and get_from_scope pair in the generated instruction stream.

Bytecodes Before:
...
[  28] resolve_scope      dst:loc10, scope:loc4, var:1, resolveType:GlobalProperty, localScopeDepth:0
[  35] get_from_scope     dst:loc7, scope:loc10, var:1, getPutInfo:2048&lt;...&gt;, localScopeDepth:0, offset:0
...

Bytecodes After:
...
[  28] resolve_and_get_from_scope dst:loc7, scope:loc4, resolvedScope:loc10, var:1, resolveType:GlobalProperty, getPutInfo:2048&lt;...&gt;, localScopeDepth:0, offset:0
...

Reviewed by NOBODY (OOPS!).

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeLivenessAnalysis.cpp:
(JSC::tmpLivenessForCheckpoint):
* Source/JavaScriptCore/bytecode/BytecodeOperandsForCheckpoint.h:
(JSC::valueProfileForImpl):
(JSC::valueProfileFor):
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finishCreation):
(JSC::CodeBlock::finalizeLLIntInlineCaches):
(JSC::CodeBlock::notifyLexicalBindingUpdate):
(JSC::CodeBlock::tryGetValueProfileForBytecodeIndex):
* Source/JavaScriptCore/bytecode/CodeBlockInlines.h:
(JSC::CodeBlock::forEachValueProfile):
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.cpp:
(JSC::UnlinkedCodeBlock::allocateSharedProfiles):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::initializeDefaultParameterValuesAndSetupFunctionScopeStack):
(JSC::BytecodeGenerator::hoistSloppyModeFunctionIfNecessary):
(JSC::BytecodeGenerator::emitResolveScopeWithVarKindScope):
(JSC::BytecodeGenerator::emitResolveScopeHelper):
(JSC::BytecodeGenerator::emitResolveScope):
(JSC::BytecodeGenerator::emitGetFromScopeWithVarKindStack):
(JSC::BytecodeGenerator::emitGetFromScopeWithVarKindDirectArgument):
(JSC::BytecodeGenerator::emitGetFromScopeHelper):
(JSC::BytecodeGenerator::emitGetFromScope):
(JSC::BytecodeGenerator::emitResolveAndGetFromScope):
(JSC::BytecodeGenerator::emitLoadThisFromArrowFunctionLexicalEnvironment):
(JSC::BytecodeGenerator::emitGetGlobalPrivate): Deleted.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ResolveNode::emitBytecode):
(JSC::BaseDotNode::emitGetPropertyValue):
(JSC::BaseDotNode::emitPutProperty):
(JSC::PostfixNode::emitDot):
(JSC::PrefixNode::emitDot):
(JSC::EmptyVarExpression::emitBytecode):
(JSC::DefineFieldNode::emitBytecode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::needsDynamicLookup):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGCapabilities.cpp:
(JSC::DFG::capabilityLevel):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::operationResolveScopeHelper):
(JSC::JSC_DEFINE_JIT_OPERATION):
(JSC::operationGetFromScopeHelper):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::emit_op_resolve_scope_helper):
(JSC::JIT::emit_op_resolve_scope):
(JSC::JIT::generateOpResolveScopeThunkHelper):
(JSC::JIT::generateOpResolveScopeThunk):
(JSC::JIT::slow_op_resolve_scopeGenerator_helper):
(JSC::JIT::slow_op_resolve_scopeGenerator):
(JSC::JIT::emit_op_get_from_scope_helper):
(JSC::JIT::emit_op_get_from_scope):
(JSC::JIT::generateOpGetFromScopeThunk):
(JSC::JIT::generateOpGetFromScopeThunkHelper):
(JSC::JIT::slow_op_get_from_scopeGenerator):
(JSC::JIT::slow_op_get_from_scopeGenerator_helper):
(JSC::JIT::emit_op_resolve_and_get_from_scope):
(JSC::JIT::generateOpRGSResolveScopeThunk):
(JSC::JIT::generateOpRGSGetFromScopeThunk):
(JSC::JIT::slow_op_rgs_resolve_scopeGenerator):
(JSC::JIT::slow_op_rgs_get_from_scopeGenerator):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::scopeGetPropertyHelper):
(JSC::LLInt::llintSlowPathGetFromScopeHelper):
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
(JSC::LLInt::handleResolveAndGetFromScopeCheckpoint):
(JSC::LLInt::llint_slow_path_checkpoint_osr_exit):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::commonSlowPathResolveScopeHelper):
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/CommonSlowPaths.h:
* Source/JavaScriptCore/runtime/CommonSlowPathsInlines.h:
(JSC::CommonSlowPaths::tryCacheGetFromScopeGlobal):
</pre>

saambarati commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1044,6 +1044,55 @@
+            globalLexicalBindingEpoch: unsigned,
+        },
+        _2: { # offset 11
+             # written during linking
```
I don't think we need all these comments about offsets. Feel free to remove it from the other ops, too

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1044,6 +1044,55 @@
+    },
+    checkpoints: {
+        cp1: nil,
+        cp2: nil,
```
let's give these English names about what they're doing.

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
```diff
@@ -50,6 +50,7 @@ CallLinkInfo::CallType CallLinkInfo::callTypeFor(OpcodeID opcodeID)
     case op_call_eval:
     case op_iterator_open:
     case op_iterator_next:
+    // case op_resolve_and_get_from_scope:
```
we don't need this

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -7992,12 +8009,263 @@ void ByteCodeParser::parseBlock(unsigned limit)
             NEXT_OPCODE(op_get_from_scope);
         }
 
+        case op_resolve_and_get_from_scope: {
```
let's try to share code with resolve_scope and get_from_scope

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/jit/JITOperations.h:
```diff
@@ -299,6 +299,8 @@ JSC_DECLARE_JIT_OPERATION(operationSwitchImmWithUnknownKeyType, char*, (VM*, Enc
 JSC_DECLARE_JIT_OPERATION(operationResolveScopeForBaseline, EncodedJSValue, (JSGlobalObject*, const JSInstruction* bytecodePC));
 JSC_DECLARE_JIT_OPERATION(operationGetFromScope, EncodedJSValue, (JSGlobalObject*, const JSInstruction* bytecodePC));
+JSC_DECLARE_JIT_OPERATION(operationResolveAndGetFromScope, EncodedJSValue, (JSGlobalObject*, const JSInstruction* bytecodePC));
+JSC_DECLARE_JIT_OPERATION(operationResolveAndGetFromScope2, EncodedJSValue, (JSGlobalObject*, const JSInstruction* bytecodePC));
```
let's use better names

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1890,12 +1918,123 @@ MacroAssemblerCodeRef<JITThunkPtrTag> JIT::slow_op_get_from_scopeGenerator(VM& v
     return FINALIZE_THUNK(patchBuffer, JITThunkPtrTag, "Baseline: slow_op_get_from_scope");
 }
 
+void JIT::emit_op_resolve_and_get_from_scope(const JSInstruction* currentInstruction)
```
Can we share more code here with resolve_scope and get_from_scope?

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
```diff
@@ -2168,7 +2172,45 @@ LLINT_SLOW_PATH_DECL(slow_path_get_from_scope)
+    auto bytecode = pc->as<OpResolveAndGetFromScope>();
+    auto& metadata = bytecode.metadata(codeBlock);
+    const Identifier& ident = codeBlock->identifier(bytecode.m_var);
+    JSObject* scope = jsCast<JSObject*>(getNonConstantOperand(callFrame, bytecode.m_resolvedScope));
```
This helper is only called after we've already resolved our scope. Might be worth giving it a name that better indicates that.

saambarati commented on Source/JavaScriptCore/bytecode/Opcode.h:
```diff
@@ -123,6 +123,7 @@ static constexpr unsigned bitWidthForMaxBytecodeStructLength = WTF::getMSBSetCon
     macro(OpCallEval) \
     macro(OpConstruct) \
     macro(OpGetFromScope) \
+    macro(OpResolveAndGetFromScope) \
```
I think this is slightly wrong. We only want to give out the value profile if checkpoint == 1. I'd do it how iterator_next does it.

hyjorc1 replied:
Updated

saambarati reviewed (commented):
I'd try to work on sharing a bit more code between the implementations, and getting this patch into a state where it can be reviewed a bit more easily. Right now, it contains a lot of duplicate code.

saambarati commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1044,6 +1044,55 @@
+        var: unsigned,
+        # $begin: :private,
+        resolveType: ResolveType,
+        getPutInfo: GetPutInfo,
```
Why do we need both of these?

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1044,6 +1044,55 @@
+        },
+        operand: uintptr_t,
+        profile: ValueProfile,
+        resolveType: ResolveType,
```
why do we need this and getPutInfo in metadata?

hyjorc1 replied:
I think they are actively used on runtime to update the profiling info. (JITOperations.cpp, JITPropertyAccess.cpp, LLIntSlowPaths.cpp, and CommonSlowPaths.cpp)

saambarati replied:
We should refactor that code so we don't need these duplicate fields. GetPutInfo has a field in it that is ResolvedType. It's entirely duplicative to have that field stored twice.

justinmichaud commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -467,6 +467,53 @@ bool CodeBlock::finishCreation(VM& vm, ScriptExecutable* ownerExecutable, Unlink
 
+    auto linkOpResolveScope = [&](auto& bytecode, auto& metadata) {
+        const Identifier& ident = identifier(bytecode.m_var);
+        RELEASE_ASSERT(bytecode.m_resolveType != ResolvedClosureVar);
```
Do we know why this is here?

hyjorc1 replied:
This is extracted from the resolve scope switch case.

justinmichaud commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -1134,16 +1146,22 @@ JSC_DEFINE_COMMON_SLOW_PATH(slow_path_resolve_scope_for_hoisting_func_decl_in_ev
     JSScope* scope = callFrame->uncheckedR(bytecode.m_scope).Register::scope();
     JSObject* resolvedScope = JSScope::resolve(globalObject, scope, ident);
+    bool isOpResolveScope = std::is_same<Bytecode, OpResolveScope>::value;
+    VirtualRegister dst = isOpResolveScope ? pc->as<Bytecode>().m_dst : pc->as<OpResolveAndGetFromScope>().m_resolvedScope;
```
Maybe a switch would be better, since it would be easier to grep for? Or perhaps the dst can be passed in as a parameter?

hyjorc1 replied:
Updated

justinmichaud commented on Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
```diff
@@ -1134,16 +1146,22 @@ JSC_DEFINE_COMMON_SLOW_PATH(slow_path_resolve_scope_for_hoisting_func_decl_in_ev
     // Proxy can throw an error here, e.g. Proxy in with statement's @unscopables.
-    CHECK_EXCEPTION();
+    CallFrame* returnSecond = isOpResolveScope ? callFrame : nullptr;
+    // Use the second field of SlowPathReturnType as a flag to indicate whether
```
This seems sub-optimal? Maybe @Constellation has some thoughts? Do iterators have to do something similar?

justinmichaud commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -2508,6 +2508,33 @@ RegisterID* BytecodeGenerator::emitResolveScope(RegisterID* dst, const Variable&
+    ASSERT(scope);
+
+#if USE(JSVALUE64)
+    auto validResolveAndGetFromScopePair = [&]() {
```
Since we only call this once, can this just be a bool?

hyjorc1 replied:
Updated

justinmichaud commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -7830,174 +8013,22 @@ void ByteCodeParser::parseBlock(unsigned limit)
+            auto bytecode = currentInstruction->as<OpResolveAndGetFromScope>();
+            auto setResolvedScope = [&] (VirtualRegister resolvedScope, Node* value) {
+                set(resolvedScope, value);
+                set(Operand::tmp(bytecode.tmpResolvedScope), value);
```
Why do we set two things here?

hyjorc1 replied:
tmpResolvedScope is used to store the resolved scope for OSR exit.

justinmichaud reviewed (commented):
I have a few questions about the code, maybe we can brainstorm later with Yusuke?

saambarati commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1044,6 +1044,53 @@
+op :resolve_and_get_from_scope,
+    args: {
+        dst: VirtualRegister,
+        scope: VirtualRegister,
```
Why do we need this operand anymore? The whole point of switching to tmps is so that we don't define a scope anymore. And we can do this if we only convert places in the bytecode generator where we can prove that the result of the resolve_scope is only used by the get_from_scope.

hyjorc1 replied:
I think we cannot remove this `scope` operand. Here, the `scope` is the input for resolve_scope.

Before:
```
[  28] resolve_scope       dst:loc10, scope:loc4, var:1, resolveType:GlobalProperty, localScopeDepth:0
[  35] get_from_scope     dst:loc7, scope:loc10, var:1, getPutInfo:2048<...>, localScopeDepth:0, offset:0
```

Please check the emitted bytecode for resolve_scope(see example above), which needs a `dst` operand and a `scope` operand. Since the destination of resolve_scope might be the input `scope` for get_from_scope only, we can introduce our new opcode resolve_and_get_from_scope by getting rid of the `dst` of resolve_scope (`dst:loc10`, in the example) and `scope` of get_from_scope (`scope:loc10`, in the example).

After:
```
[  28] resolve_and_get_from_scope dst:loc7, scope:loc4, resolvedScope:loc10, var:1, resolveType:GlobalProperty, getPutInfo:2048<...>, localScopeDepth:0, offset:0
```

However, we still need `scope:loc4` to get the resolved scope and `dst:loc7`as the destination of get_from_scope. That's why we cannot remove this `scope` operand in resolve_and_get_from_scope unless we have an approach to resolve scope without the input `scope`.

Previously, in this patch we have three operands (`dst`, `resolvedScope`, and `scope`), where `resolvedScope` is removed in current version since we only consider the cases that `resolvedScope` is not used by other emitted instructions.

saambarati replied:
ignore me, I wasn't thinking right. This is needed as an input

saambarati commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -270,10 +270,9 @@ RegisterID* ResolveNode::emitBytecode(BytecodeGenerator& generator, RegisterID*
     RegisterID* finalDest = generator.finalDestination(dst);
     RefPtr<RegisterID> uncheckedResult = generator.newTemporary();
-    generator.emitGetFromScope(uncheckedResult.get(), scope.get(), var, ThrowIfNotFound);
+    generator.emitResolveAndGetFromScope(uncheckedResult.get(), dst, var, ThrowIfNotFound);
```
Why are we padding `dst` as the resolved scope here? That looks quite wrong.

hyjorc1 replied:
I think the logic is correct. But, the function name `emitResolveAndGetFromScope` is misleading and it should be `tryEmitResolveAndGetFromScope`. If you check the implementation of this function:
```
RegisterID* BytecodeGenerator::emitResolveAndGetFromScope(RegisterID* dst, RegisterID* resolvedScope, const Variable& variable, ResolveMode resolveMode, bool isPrivate)
{
    switch (variable.offset().kind()) {
    case VarKind::Stack: {
        ASSERT(!isPrivate);
        UNUSED_VARIABLE(isPrivate);
        return emitGetFromScopeWithVarKindStack(dst, variable);
    }
    case VarKind::DirectArgument:
        return emitGetFromScopeWithVarKindDirectArgument(dst, argumentsRegister(), variable);
    case VarKind::Scope:
        return emitGetFromScopeHelper(dst, emitResolveScopeWithVarKindScope(variable), variable, resolveMode);
    case VarKind::Invalid: {
#if USE(JSVALUE64)
        if (!variable.offset().isScope()) {
            OpResolveAndGetFromScope::emit(
                this,
                kill(dst),
                scopeRegister(),
                addConstant(variable.ident()),
                GetPutInfo(resolveMode, resolveType(), InitializationMode::NotInitialization, ecmaMode()),
                localScopeDepth(),
                variable.offset().isScope() ? variable.offset().scopeOffset().offset() : 0);
            return dst;
        }
#endif
        return emitGetFromScopeHelper(dst, emitResolveScopeHelper(resolvedScope, variable), variable, resolveMode);
    }
    }

    RELEASE_ASSERT_NOT_REACHED();
    return nullptr;
}
```
which still need to handle the non-ResolveAndGetFromScope cases. Here, the `dst` is passed to the function as the resolvedScope for clause:
```
return emitGetFromScopeHelper(dst, emitResolveScopeHelper(resolvedScope, variable), variable, resolveMode);
```

hyjorc1 replied:
I believe the similar situation happens in both `emitResolveScope` and `emitGetFromScope`. And they should be `tryEmitResolveScope` and `tryEmitGetFromScope` instead.

saambarati replied:
They should not be prefixed with `try`. Whether or not we actually emit the opcodes, we're logically resolving and getting from a scope. I don't see why we need to pass in the `dst` as the result of the resolve_scope. If we need the scope for later operations, we shouldn't be emitting resolve_and_get_from_scope combo opcode.

saambarati commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
```diff
@@ -846,8 +845,14 @@ namespace JSC {
+        RegisterID* emitGetFromScopeWithVarKindDirectArgument(RegisterID* dst, RegisterID* scope, const Variable&);
+        RegisterID* emitGetFromScopeHelper(RegisterID* dst, RegisterID* scope, const Variable&, ResolveMode);
         RegisterID* emitGetFromScope(RegisterID* dst, RegisterID* scope, const Variable&, ResolveMode);
+        RegisterID* emitResolveAndGetFromScope(RegisterID* dst, RegisterID* resolvedScope, const Variable&, ResolveMode, bool isPrivate = false);
```
I don't see a reason why we need a resolvedScope parameter here. The one call site that's passing something other than nullptr is looks wrong to me. I think it's actually better if the contract of this API is that we don't define a scope, so we can remove the operand from the bytecode

hyjorc1 replied:
Yes, we need `resolvedScope` for handling non-ResolveAndGetFromScope cases introduced in `emitResolveAndGetFromScope`.

saambarati commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4279,12 +4280,18 @@ GetByOffsetMethod ByteCodeParser::promoteToConstant(GetByOffsetMethod method)
+    bool isResolveScope = false;
+    if (opcode == op_resolve_scope)
+        isResolveScope = true;
+    if (opcode == op_resolve_and_get_from_scope && m_currentIndex.checkpoint() == OpResolveAndGetFromScope::resolveScope)
```
nit: this can be an `else if`

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -5456,6 +5463,284 @@ void ByteCodeParser::parseBlock(unsigned limit)
+
+        if (needsDynamicLookup(resolveType, op_resolve_scope)) {
+            unsigned identifierNumber = m_inlineStackTop->m_identifierRemap[bytecode.m_var];
+            set(dst, addToGraph(ResolveScope, OpInfo(identifierNumber), get(bytecode.m_scope)));
```
This doesn't look right. It's not putting the result in `dst`

saambarati replied:
Also, this would be a compile error once we get rid of that `scope` field we no longer need from the bytecode

hyjorc1 replied:
> This doesn't look right. It's not putting the result in `dst`

I think this is correct. Here, the `dst` is passed to the helper function `resolveScopeHelper `, which is extracted from `case op_resolve_scope:` and it is for both op_resolve_scope and op_resolve_and_get_from_scope. Each case will pass their own `dst`, which is the a destination register or a Tmp variable for op_resolve_scope or op_resolve_and_get_from_scope respectively. 

> Also, this would be a compile error once we get rid of that scope field we no longer need from the bytecode

I don't think we can remove the `scope` operand or field.

saambarati commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1437,11 +1437,44 @@ void JIT::emitSlow_op_has_private_brand(const JSInstruction*, Vector<SlowCaseEnt
-void JIT::emit_op_resolve_scope(const JSInstruction* currentInstruction)
-{
-    auto bytecode = currentInstruction->as<OpResolveScope>();
+#define RESOLVE_SCOPE_CODE_REF(thunk)                                              \
```
Can this be a function instead of a macro?

hyjorc1 replied:
The passed thunk function is a template function which cannot be passed as a function pointer. If we need to use function, then we have to create two functions for OpGetFromScope and OpResolveScope each.

saambarati commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1660,16 +1691,51 @@ MacroAssemblerCodeRef<JITThunkPtrTag> JIT::slow_op_resolve_scopeGenerator(VM& vm
+    return slow_op_resolve_scopeGenerator_helper<OpResolveScope>(vm, operationResolveScopeForBaseline);
+}
+
+#define GET_FROM_SCOPE_CODE_REF(thunk)                                             \
```
ditto, can this be a function instead of a macro?

hyjorc1 replied:
ditto, are two functions better than one macro?

saambarati commented on Source/JavaScriptCore/jit/JIT.h:
```diff
@@ -668,11 +674,25 @@ namespace JSC {
+        template <ResolveType>
+        static MacroAssemblerCodeRef<JITThunkPtrTag> generateOpRGSResolveScopeThunk(VM&);
+        template <ResolveType>
+        static MacroAssemblerCodeRef<JITThunkPtrTag> generateOpRGSGetFromScopeThunk(VM&);
```
nit: Let's not abbreviate this name this way. Write out the full thing

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
```diff
@@ -1890,12 +1960,40 @@ MacroAssemblerCodeRef<JITThunkPtrTag> JIT::slow_op_get_from_scopeGenerator(VM& v
 
+void JIT::emit_op_resolve_and_get_from_scope(const JSInstruction* currentInstruction)
+{
+    emit_op_resolve_scope_helper<OpResolveAndGetFromScope>(currentInstruction, currentInstruction->as<OpResolveAndGetFromScope>().m_dst);
+    emit_op_get_from_scope_helper<OpResolveAndGetFromScope>(currentInstruction, currentInstruction->as<OpResolveAndGetFromScope>().m_dst);
```
We should refactor this in such a way we don't need to write out the intermediate resolve_scope result to the stack. There is no need to do that, so we should avoid it

hyjorc1 replied:
The current implementation stores the intermediate resolve_scope result to the `m_dst` of op_resovle_and_get_from_scope. So, it doesn't need extra space in the stack to keep it. Are you proposing that we should store it to an GPR instead? Could that perform better performance in read and write?

saambarati commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "DFGByteCodeParser.h"
+#include "interpreter/CallFrame.h"
```
This can just be `#include "CallFrame.h"`

hyjorc1 replied:
Updated

saambarati commented on Source/JavaScriptCore/llint/LLIntSlowPaths.h:
```diff
@@ -137,6 +137,7 @@ LLINT_SLOW_PATH_HIDDEN_DECL(slow_path_handle_traps);
 LLINT_SLOW_PATH_HIDDEN_DECL(slow_path_debug);
 LLINT_SLOW_PATH_HIDDEN_DECL(slow_path_handle_exception);
 LLINT_SLOW_PATH_HIDDEN_DECL(slow_path_get_from_scope);
+LLINT_SLOW_PATH_HIDDEN_DECL(slow_path_rgs_get_from_scope);
```
nit: same thing about writing out the full name

hyjorc1 replied:
Updated

hyjorc1 committed:
[JSC] Add a resolve_and_get_from_scope bytecode opcode
https://bugs.webkit.org/show_bug.cgi?id=237836

Reviewed by NOBODY (OOPS!).

In generated bytecodes, resolve_scope followed by get_from_scope is common.
This patch introduces a new bytecode opcode resolve_and_get_from_scope by
replacing the resolve_scope and get_from_scope pair in the generated instruction stream.

Bytecodes Before:
...
[  28] resolve_scope      dst:loc10, scope:loc4, var:1, resolveType:GlobalProperty, localScopeDepth:0
[  35] get_from_scope     dst:loc7, scope:loc10, var:1, getPutInfo:2048<...>, localScopeDepth:0, offset:0
...

Bytecodes After:
...
[  28] resolve_and_get_from_scope dst:loc7, scope:loc4, resolvedScope:loc10, var:1, resolveType:GlobalProperty, getPutInfo:2048<...>, localScopeDepth:0, offset:0
...

Reviewed by NOBODY (OOPS!).

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeLivenessAnalysis.cpp:
(JSC::tmpLivenessForCheckpoint):
* Source/JavaScriptCore/bytecode/BytecodeOperandsForCheckpoint.h:
(JSC::valueProfileForImpl):
(JSC::valueProfileFor):
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finishCreation):
(JSC::CodeBlock::finalizeLLIntInlineCaches):
(JSC::CodeBlock::notifyLexicalBindingUpdate):
(JSC::CodeBlock::tryGetValueProfileForBytecodeIndex):
* Source/JavaScriptCore/bytecode/CodeBlockInlines.h:
(JSC::CodeBlock::forEachValueProfile):
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.cpp:
(JSC::UnlinkedCodeBlock::allocateSharedProfiles):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::initializeDefaultParameterValuesAndSetupFunctionScopeStack):
(JSC::BytecodeGenerator::hoistSloppyModeFunctionIfNecessary):
(JSC::BytecodeGenerator::emitResolveScopeWithVarKindScope):
(JSC::BytecodeGenerator::emitResolveScopeHelper):
(JSC::BytecodeGenerator::emitResolveScope):
(JSC::BytecodeGenerator::emitGetFromScopeWithVarKindStack):
(JSC::BytecodeGenerator::emitGetFromScopeWithVarKindDirectArgument):
(JSC::BytecodeGenerator::emitGetFromScopeHelper):
(JSC::BytecodeGenerator::emitGetFromScope):
(JSC::BytecodeGenerator::emitResolveAndGetFromScope):
(JSC::BytecodeGenerator::emitLoadThisFromArrowFunctionLexicalEnvironment):
(JSC::BytecodeGenerator::emitGetGlobalPrivate): Deleted.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ResolveNode::emitBytecode):
(JSC::BaseDotNode::emitGetPropertyValue):
(JSC::BaseDotNode::emitPutProperty):
(JSC::PostfixNode::emitDot):
(JSC::PrefixNode::emitDot):
(JSC::EmptyVarExpression::emitBytecode):
(JSC::DefineFieldNode::emitBytecode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::needsDynamicLookup):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGCapabilities.cpp:
(JSC::DFG::capabilityLevel):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::operationResolveScopeHelper):
(JSC::JSC_DEFINE_JIT_OPERATION):
(JSC::operationGetFromScopeHelper):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::emit_op_resolve_scope_helper):
(JSC::JIT::emit_op_resolve_scope):
(JSC::JIT::generateOpResolveScopeThunkHelper):
(JSC::JIT::generateOpResolveScopeThunk):
(JSC::JIT::slow_op_resolve_scopeGenerator_helper):
(JSC::JIT::slow_op_resolve_scopeGenerator):
(JSC::JIT::emit_op_get_from_scope_helper):
(JSC::JIT::emit_op_get_from_scope):
(JSC::JIT::generateOpGetFromScopeThunk):
(JSC::JIT::generateOpGetFromScopeThunkHelper):
(JSC::JIT::slow_op_get_from_scopeGenerator):
(JSC::JIT::slow_op_get_from_scopeGenerator_helper):
(JSC::JIT::emit_op_resolve_and_get_from_scope):
(JSC::JIT::generateOpRGSResolveScopeThunk):
(JSC::JIT::generateOpRGSGetFromScopeThunk):
(JSC::JIT::slow_op_rgs_resolve_scopeGenerator):
(JSC::JIT::slow_op_rgs_get_from_scopeGenerator):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::scopeGetPropertyHelper):
(JSC::LLInt::llintSlowPathGetFromScopeHelper):
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
(JSC::LLInt::handleResolveAndGetFromScopeCheckpoint):
(JSC::LLInt::llint_slow_path_checkpoint_osr_exit):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::commonSlowPathResolveScopeHelper):
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/CommonSlowPaths.h:
* Source/JavaScriptCore/runtime/CommonSlowPathsInlines.h:
(JSC::CommonSlowPaths::tryCacheGetFromScopeGlobal):
